lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchersyncer
+
+import (
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+const (
+	maxUpdatesToConsolidate = 1000
+)
+
+// ResourceType groups together the watch and conversion information for a
+// specific resource type.
+type ResourceType struct {
+	// The ListInterface used to perform a watch on this resource type.
+	ListInterface model.ListInterface
+
+	// An update processor used to convert the update event prior to it being sent in
+	// the Syncer.
+	UpdateProcessor SyncerUpdateProcessor
+}
+
+// SyncerUpdateProcessor is used to convert a Watch update into one or more additional
+// Syncer updates.
+type SyncerUpdateProcessor interface {
+	// Process is called to process a watch update.  The processor may convert this
+	// to zero or more updates.  The processor may use these calls to maintain a local cache
+	// if required.  It is safe for the processor to send multiple duplicate adds or deletes
+	// since the WatcherSyncer maintains it's own cache and will swallow duplicates.
+	// A KVPair with a nil value indicates a delete.  A non nil value indicates an add/modified.
+	// The processor may respond with any number of adds or deletes.
+	Process(*model.KVPair) ([]*model.KVPair, error)
+
+	// SyncStarting is called when syncer is starting a full sync for the associated resource
+	// type.  That means it is first going to list current resources and then watch for any updates.
+	// If the processor maintains a private internal cache, then the cache should be cleared at
+	// this point since the cache will be re-populated from the sync.
+	SyncStarting()
```

Comment:
another naming nit - what about `OnSyncerStart` to better indicate this is something called when the syncer starts

Reply:
nitpick away.... :-)